doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,141 @@
+
+# Non-Goals
+
+None of the following are intended to be done with this design:
+
+* Export our system table logs as prometheus (yet) -- this design covers only the "metrics import" direction.
+
+* Expose any `CREATE SOURCE` mechanics for importing from arbitrary/external prometheus endpoints.
+
+* Come up with a data retention scheme other than "older than n minutes gets dropped". If this was a full-featured time series database design, we'd want to thin metrics into minute/hour/day buckets. This isn't that.
+
+# Description
+
+<!--
+    Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+    If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+By implementing this design, we'll make the following changes to materialized:
+
+1. Introduce a background task that regularly scrapes the internal metrics registry using the registry's [`gather`](https://docs.rs/prometheus/0.12.0/prometheus/struct.Registry.html#method.gather) method (currently we can use the global registry, but later we'll need a handle to the ["materialized" metrics registry][#5825]; the mechanism will remain the same).
+2. Ingest the generated metrics into a dataflow that populates metrics into tables like the following:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+
+CREATE TABLE metrics_table_meta (
+    metric TEXT,
+    metric_type TEXT,
+    help TEXT
+);
+```
+
+3. This not-quite-source polls the endpoint once a minute (via the configured the `poll_interval` duration), and retains metric readings for a number of minutes (strawman proposal: 5 minutes) to make it possible for users to visualize changes in the metrics as they would with a "full" metrics pipeline.
+
+## Possible work after this proposal is implemented
+
+* As noted above, after [#5825] is implemented, we'll have to adjust the import task outlined above to receive & use the new registry.
+
+* Once we have a way to import our process's own prometheus metrics, one could introduce a way (a full-fledged `SOURCE`) to read from arbitrary prometheus endpoints. This is explicitly a non-goal right now, and doesn't really align with our product direction right now... but it would be a fun skunkworks project.
+
+# Alternatives
+
+<!-- // Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen -->
+
+There are two major things that we could do differently to fulfill the brief, and a series of things that we could do to polish this design. I'll summarize the former under "Alternative approaches" and the latter under "Alternative details".
+
+## Alternative approaches
+
+### Create an internal metrics reporting channel instead (feeding metrics directly)
```

Comment:
We already have the timely logs which sound very similar to what you're describing here -- what is the difference between this approach and that?

Reply:
The main (only real) difference is the interface at the point of metric value emission - currently, things using the prometheus interface have an easier time, e.g. tracking histograms (which we have to do [faiiirly manually in timely logging code](https://github.com/MaterializeInc/materialize/blob/ebed079124da5fdcc50a0e97be947b4c39fac3c6/src/dataflow/src/logging/timely.rs#L331-L338)). That, together with the constraints mentioned in the tracking issue (https://github.com/MaterializeInc/materialize/issues/1696#issuecomment-587283328), my understanding is we will want to retain prometheus as an external and internal interface (external to ensure that materialized run can be measured right by our users, internal to ensure the metrics we track actually make it to the external interface).

So I think we'd have two options for getting the same values into prometheus's external interface & our internal log tables: 
* importing the metric values from the prometheus registry via some public interface (the design proposed here), or
* doing the work twice at the point of metric value emission: incrementing a prometheus counter & sending an event down a channel that gets counted / processed into a log table entry (this alternative approach).

I don't much love the second option because it means we effectively do the work twice and we have to ~copy the behavior of prometheus or risk questions about which number is right (the old adage about a man with two wristwatches never knowing what time it is).

Reply:
This is also a reference to the design in #5277, which had the goal of making it easier to insert into timely logs from places in out codebase that aren't the `dataflow` crate.